include/ttmlir/Dialect/TTNN/Transforms/Optimizer.h

```diff
@@ -0,0 +1,163 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <mlir/Pass/PassRegistry.h>
+
+#ifdef GEN_PASS_DECL
+#define GEN_PASS_DECL_TTNNOPTIMIZER
+#undef GEN_PASS_DECL
+#endif // GEN_PASS_DECL
+
+//===----------------------------------------------------------------------===//
+// TTNNOptimizer
+//===----------------------------------------------------------------------===//
+#ifdef GEN_PASS_DECL_TTNNOPTIMIZER
+struct TTNNOptimizerOptions {
+  llvm::StringMap<InputLayoutOverrideParams> overrideInputLayout =
+      llvm::StringMap<InputLayoutOverrideParams>();
+  llvm::StringMap<OutputLayoutOverrideParams> overrideOutputLayout =
+      llvm::StringMap<OutputLayoutOverrideParams>();
+  bool shardingPassEnabled = false;
+  bool reshardingEnabled = false;
+  int64_t maxLegalLayouts = 64;
+};
+std::unique_ptr<::mlir::Pass> createTTNNOptimizer();
+std::unique_ptr<::mlir::Pass> createTTNNOptimizer(TTNNOptimizerOptions options);
+#undef GEN_PASS_DECL_TTNNOPTIMIZER
+#endif // GEN_PASS_DECL_TTNNOPTIMIZER
+#ifdef GEN_PASS_DEF_TTNNOPTIMIZER
+
+namespace impl {
+std::unique_ptr<::mlir::Pass> createTTNNOptimizer();
+} // namespace impl
+
+namespace impl {
+std::unique_ptr<::mlir::Pass> createTTNNOptimizer(TTNNOptimizerOptions options);
+} // namespace impl
+namespace impl {
+
+template <typename DerivedT>
+class TTNNOptimizerBase : public ::mlir::OperationPass<::mlir::ModuleOp> {
```

Comment:
This can be trimmed down significantly see https://mlir.llvm.org/docs/PassManagement/:
`struct MyPass : PassWrapper<MyPass, OperationPass<ModuleOp>> {
  MyPass() = default;
  MyPass(const MyPass &) {}

  ...

  // Specify any options.
  Option<bool> option{
      *this, "example-option",
      llvm::cl::desc("An example option"), llvm::cl::init(true)};
  ListOption<int64_t> listOption{
      *this, "example-list",
      llvm::cl::desc("An example list option")};

  // Specify any statistics.
  Statistic statistic{this, "example-statistic", "An example statistic"};
};

/// Expose this pass to the outside world.
std::unique_ptr<Pass> foo::createMyPass() {
  return std::make_unique<MyPass>();
}

/// Register this pass.
void foo::registerMyPass() {
  PassRegistration<MyPass>();
}`

We dont need the baseclass anymore. We can define OptimizerPass here directly and just implement runOnOperation in Optimizer.cpp. But its ok not to do it now.

Reply:
Yeah we can leave as is for now.